common/environment/environment.cpp

```diff
@@ -362,7 +362,7 @@ class CSdsSubscription : implements ISDSSubscription, public CInterface
     CSdsSubscription()
     {
         m_constEnvUpdated = false;
-        Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory();
+        Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory(false);
```

Comment:
not new but a bit odd that that this class (CSdsSubscription) - which is only ever created by CEnvironmentFactory, asks for the factory here.
It would be more obvious if the factory was passed into the CSdsSubscription constructor.

Reply:
If I pass the factory 'envFactory' into the CSdsSubscription constructor and call envFactory->subscribe() directly inside the constructor , the factory CEnvironmentFactory has to be fully declared before the CSdsSubscription. I may have to add a new .h file and move many class declarations from environment.cpp to the .h file. @jakesmith do you want this big change?

Reply:
> the factory CEnvironmentFactory has to be fully declared before the CSdsSubscription.

It does not need a CEnvironmentFactory, it is dealing with a IEnvironmentFactory, so it does not need to know a CEnvironmentFactory.

CEnvironmentFactory::validateCache() could be:
```
    virtual void validateCache()
    {
        if (!subscription)
            subscription.setown( new CSdsSubscription(*this) );

        subscription->handleEnvironmentChange();
    }
```
..and the CSdsSubscription constructor could be:
```
    CSdsSubscription(IEnvironmentFactory &_factory) : factory(_factory)
```